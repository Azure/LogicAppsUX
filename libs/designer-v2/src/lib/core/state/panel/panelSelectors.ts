import type { RootState } from '../../store';
import { createSelector } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

const getPanelState = (state: RootState) => state.panel;

export const useConnectionPanelSelectedNodeIds = () =>
  useSelector(createSelector(getPanelState, (state) => state.connectionContent.selectedNodeIds));

export const useCurrentPanelMode = () => useSelector(createSelector(getPanelState, (state) => state.currentPanelMode));

export const useDiscoveryPanelSelectedOperationGroupId = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.selectedOperationGroupId));

export const useDiscoveryPanelSelectedOperationId = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.selectedOperationId));

export const useDiscoveryPanelSelectedNodeIds = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.selectedNodeIds));

export const useDiscoveryPanelIsAddingTrigger = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.isAddingTrigger));

export const useDiscoveryPanelIsParallelBranch = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.isParallelBranch));

export const useDiscoveryPanelRelationshipIds = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.relationshipIds));

export const useDiscoveryPanelFavoriteOperations = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.favoriteOperations));

export const useDiscoveryPanelSelectedBrowseCategory = () =>
  useSelector(createSelector(getPanelState, (state) => state.discoveryContent.selectedBrowseCategory));

export const useDiscoveryPanelIsOperationFavorited = (connectorId: string, operationId?: string) =>
  useDiscoveryPanelFavoriteOperations().some((favorite) => favorite.connectorId === connectorId && favorite.operationId === operationId);

export const useErrorsPanelSelectedTabId = () => useSelector(createSelector(getPanelState, (state) => state.errorContent.selectedTabId));

export const useFocusReturnElementId = () => useSelector(createSelector(getPanelState, (state) => state.focusReturnElementId));

export const useIsCreatingConnection = () =>
  useSelector(createSelector(getPanelState, (state) => state.connectionContent.isCreatingConnection));

export const useIsPanelCollapsed = () => useSelector(createSelector(getPanelState, (state) => state.isCollapsed));

export const useIsPanelLoading = () => useSelector(createSelector(getPanelState, (state) => state.isLoading));

export const useIsNodePinnedToOperationPanel = (nodeId: string) =>
  useSelector(
    createSelector(getPanelState, (state) => {
      return (
        (state.operationContent.alternateSelectedNode?.nodeId ?? '') === nodeId &&
        (state.operationContent.alternateSelectedNode?.persistence ?? '') === 'pinned'
      );
    })
  );

export const useIsAlternateNodePinned = () =>
  useSelector(createSelector(getPanelState, (state) => (state.operationContent.alternateSelectedNode?.persistence ?? '') === 'pinned'));

export const useIsNodeSelectedInOperationPanel = (nodeId: string) =>
  useSelector(createSelector(getPanelState, (state) => (state.operationContent.selectedNodeId ?? '') === nodeId));

export const useIsPanelInPinnedViewMode = (): boolean => {
  const selectedNodeId = useOperationPanelSelectedNodeId();
  const alternateSelectedNode = useOperationAlternateSelectedNodeId();
  return !!(selectedNodeId && alternateSelectedNode && alternateSelectedNode !== selectedNodeId);
};

export const useOperationAlternateSelectedNodeId = () =>
  useSelector(createSelector(getPanelState, (state) => state.operationContent.alternateSelectedNode?.nodeId ?? ''));

export const useOperationAlternateSelectedNode = () =>
  useSelector(createSelector(getPanelState, (state) => state.operationContent.alternateSelectedNode ?? {}));

export const useOperationPanelAlternateNodeActiveTabId = () =>
  useSelector(createSelector(getPanelState, (state) => state.operationContent.alternateSelectedNode?.activeTabId));

export const useOperationPanelSelectedNodeId = () =>
  useSelector(createSelector(getPanelState, (state) => state.operationContent?.selectedNodeId ?? ''));

export const useOperationPanelSelectedNodeActiveTabId = () =>
  useSelector(createSelector(getPanelState, (state) => state.operationContent.selectedNodeActiveTabId));

export const usePanelLocation = () => useSelector(createSelector(getPanelState, (state) => state.location));

export const usePreviousPanelMode = () => useSelector(createSelector(getPanelState, (state) => state.previousPanelMode));

export const useIsAddingAgentTool = () => useSelector(createSelector(getPanelState, (state) => state.discoveryContent.isAddingAgentTool));

export const useIsRunHistoryCollapsed = () => useSelector(createSelector(getPanelState, (state) => state.runHistoryCollapsed));
