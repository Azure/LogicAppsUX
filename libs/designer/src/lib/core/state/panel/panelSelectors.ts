import type { RootState } from '../../store';
import type { PanelState } from './panelInterfaces';
import { createSelector } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

const getPanelState = (state: RootState) => state.panel;

export const useIsPanelCollapsed = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.collapsed));

export const useCurrentPanelModePanelMode = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.currentState));

export const useIsNodeSearchMode = () =>
  useSelector(createSelector(getPanelState, (state: PanelState) => state.currentState === 'NodeSearch'));

export const useIsAddingTrigger = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.addingTrigger));

export const useRelationshipIds = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.relationshipIds));

export const useIsParallelBranch = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.isParallelBranch));

export const useSelectedSearchOperationGroupId = () =>
  useSelector(createSelector(getPanelState, (state: PanelState) => state.selectedOperationGroupId));

export const useSelectedSearchOperationId = () =>
  useSelector(createSelector(getPanelState, (state: PanelState) => state.selectedOperationId));

export const useSelectedNodeId = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.selectedNode));

export const useIsNodeSelected = (nodeId: string) =>
  useSelector(createSelector(getPanelState, (state: PanelState) => state.selectedNode === nodeId));

export const useRegisteredPanelTabs = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.registeredTabs));

export const usePanelTabByName = (tabName: string) =>
  useSelector(createSelector(getPanelState, (state: PanelState) => state.registeredTabs[tabName]));

export const useVisiblePanelTabs = () =>
  useSelector(
    createSelector(getPanelState, (state: PanelState) => {
      const visibleTabs = Object.values(state.registeredTabs).filter((tab) => tab.visible !== false);
      return visibleTabs.sort((a, b) => a.order - b.order);
    })
  );

export const useSelectedPanelTabName = () => useSelector(createSelector(getPanelState, (state: PanelState) => state.selectedTabName));

export const useSelectedPanelTab = () => {
  const selectedPanelTabName = useSelectedPanelTabName();
  return useSelector(createSelector(getPanelState, (state: PanelState) => state.registeredTabs[selectedPanelTabName ?? '']));
};
