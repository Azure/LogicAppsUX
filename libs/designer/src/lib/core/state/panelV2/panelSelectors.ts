import type { RootState } from '../../store';
import { createSelector } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

const getPanelState = (state: RootState) => state.panelV2;

export const usePinnedNodeId = () => useSelector(createSelector(getPanelState, (state) => state.operationContent.pinnedNodeId ?? ''));

export const useIsNodePinned = (nodeId: string) => usePinnedNodeId() === nodeId;
