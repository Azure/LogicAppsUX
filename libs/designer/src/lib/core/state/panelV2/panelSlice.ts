import { PanelLocation } from '@microsoft/designer-ui';
import { cleanConnectorId, LogEntryLevel, LoggerService } from '@microsoft/logic-apps-shared';
import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice, isAnyOf } from '@reduxjs/toolkit';
import { resetWorkflowState } from '../global';
import {
  changePanelNode as panelV1ChangePanelNode,
  clearPanel as panelV1ClearPanel,
  collapsePanel as panelV1CollapsePanel,
  expandDiscoveryPanel as panelV1ExpandDiscoveryPanel,
  expandPanel as panelV1ExpandPanel,
  openPanel as panelV1OpenPanel,
  selectErrorsPanelTab as panelV1SelectErrorsPanelTab,
  selectOperationGroupId as panelV1SelectOperationGroupId,
  selectOperationId as panelV1SelectOperationId,
  selectPanelTab as panelV1SelectPanelTab,
  setIsCreatingConnection as panelV1SetIsCreatingConnection,
  setIsPanelLoading as panelV1SetIsPanelLoading,
  setSelectedNodeId as panelV1SetSelectedNodeId,
  setSelectedNodeIds as panelV1SetSelectedNodeIds,
  updatePanelLocation as panelV1UpdatePanelLocation,
} from '../panel/panelSlice';
import type {
  ConnectionPanelContentState,
  DiscoveryPanelContentState,
  ErrorPanelContentState,
  NodeSearchPanelContentState,
  OperationPanelContentState,
  PanelMode,
  PanelState,
  RelationshipIds,
  WorkflowParametersPanelContentState,
} from './panelTypes';

const getInitialConnectionContentState = (): ConnectionPanelContentState => ({
  isCreatingConnection: false,
  panelMode: 'Connection',
  selectedNodeIds: [],
});

const getInitialDiscoveryContentState = (): DiscoveryPanelContentState => ({
  isAddingTrigger: false,
  isParallelBranch: false,
  panelMode: 'Discovery',
  relationshipIds: {
    graphId: 'root',
  },
  selectedNodeIds: [],
  selectedOperationGroupId: '',
  selectedOperationId: '',
});

const getInitialErrorContentState = (): ErrorPanelContentState => ({
  panelMode: 'Error',
});

const getInitialNodeSearchContentState = (): NodeSearchPanelContentState => ({
  panelMode: 'NodeSearch',
});

const getInitialOperationContentState = (): OperationPanelContentState => ({
  panelMode: 'Operation',
  pinnedNodeId: undefined,
  pinnedNodeActiveTabId: undefined,
  selectedNodeId: undefined,
  selectedNodeActiveTabId: undefined,
});

const getInitialWorkflowParametersContentState = (): WorkflowParametersPanelContentState => ({
  panelMode: 'WorkflowParameters',
});

const initialState: PanelState = {
  connectionContent: getInitialConnectionContentState(),
  currentPanelMode: 'Operation',
  discoveryContent: getInitialDiscoveryContentState(),
  errorContent: getInitialErrorContentState(),
  isCollapsed: true,
  isLoading: false,
  location: PanelLocation.Right,
  nodeSearchContent: getInitialNodeSearchContentState(),
  operationContent: getInitialOperationContentState(),
  previousPanelMode: undefined,
  workflowParametersContent: getInitialWorkflowParametersContentState(),
};

const area = 'Designer:PanelV2 Slice';

export const panelSlice = createSlice({
  name: 'panelV2',
  initialState,
  reducers: {
    expandPanel: (state) => {
      state.isCollapsed = false;
    },
    collapsePanel: (state) => {
      state.isCollapsed = true;
      state.discoveryContent.selectedOperationGroupId = '';
      state.discoveryContent.isAddingTrigger = false;
    },
    clearPanel: (state, action: PayloadAction<{ clearPinnedState?: boolean } | undefined>) => {
      const { clearPinnedState } = action.payload ?? {};

      state.connectionContent = getInitialConnectionContentState();
      state.currentPanelMode = 'Operation';
      state.discoveryContent = getInitialDiscoveryContentState();
      state.errorContent = getInitialErrorContentState();
      state.nodeSearchContent = getInitialNodeSearchContentState();
      state.previousPanelMode = undefined;
      state.workflowParametersContent = getInitialWorkflowParametersContentState();

      if (clearPinnedState) {
        state.operationContent = getInitialOperationContentState();
      } else {
        state.operationContent = {
          ...getInitialOperationContentState(),
          pinnedNodeId: state.operationContent.pinnedNodeId,
          pinnedNodeActiveTabId: state.operationContent.pinnedNodeActiveTabId,
        };
      }
    },
    updatePanelLocation: (state, action: PayloadAction<PanelLocation | undefined>) => {
      if (action.payload && action.payload !== state.location) {
        state.location = action.payload;
      }
    },
    setPinnedNode: (state, action: PayloadAction<{ nodeId: string; updatePanelOpenState?: boolean }>) => {
      const { nodeId, updatePanelOpenState } = action.payload;
      const hasSelectedNode = !!state.operationContent.selectedNodeId;

      if (nodeId && !hasSelectedNode) {
        state.connectionContent.selectedNodeIds = [nodeId];
        state.discoveryContent.selectedNodeIds = [nodeId];
        state.operationContent.selectedNodeId = nodeId;
      }

      state.operationContent.pinnedNodeId = nodeId;
      state.operationContent.pinnedNodeActiveTabId = undefined;

      if (updatePanelOpenState) {
        if (nodeId) {
          state.isCollapsed = false;
          state.currentPanelMode = 'Operation';
        } else {
          state.isCollapsed = !hasSelectedNode;
        }
      }
    },
    setSelectedNodeId: (state, action: PayloadAction<string>) => {
      const selectedNodes = [action.payload];

      state.connectionContent.selectedNodeIds = selectedNodes;
      state.discoveryContent.selectedNodeIds = selectedNodes;
      state.operationContent.selectedNodeId = selectedNodes[0];
    },
    setSelectedNodeIds: (state, action: PayloadAction<string[]>) => {
      const selectedNodes = action.payload;

      state.connectionContent.selectedNodeIds = selectedNodes;
      state.discoveryContent.selectedNodeIds = selectedNodes;
      state.operationContent.selectedNodeId = selectedNodes[0];
    },
    changePanelNode: (state, action: PayloadAction<string>) => {
      const selectedNodes = [action.payload];

      clearPanel();

      state.isCollapsed = false;
      state.currentPanelMode = 'Operation';
      state.connectionContent.selectedNodeIds = selectedNodes;
      state.operationContent.selectedNodeId = selectedNodes[0];
      state.operationContent.selectedNodeActiveTabId = undefined;

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
    expandDiscoveryPanel: (
      state,
      action: PayloadAction<{
        addingTrigger?: boolean;
        focusReturnElementId?: string;
        isParallelBranch?: boolean;
        nodeId: string;
        relationshipIds: RelationshipIds;
      }>
    ) => {
      const { addingTrigger, focusReturnElementId, isParallelBranch, nodeId, relationshipIds } = action.payload;

      state.currentPanelMode = 'Discovery';
      state.focusReturnElementId = focusReturnElementId;
      state.isCollapsed = false;
      state.discoveryContent.isAddingTrigger = !!addingTrigger;
      state.discoveryContent.isParallelBranch = isParallelBranch ?? false;
      state.discoveryContent.relationshipIds = relationshipIds;
      state.discoveryContent.selectedNodeIds = [nodeId];

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
    selectOperationGroupId: (state, action: PayloadAction<string>) => {
      state.discoveryContent.selectedOperationGroupId = cleanConnectorId(action.payload);

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
    selectOperationId: (state, action: PayloadAction<string>) => {
      state.discoveryContent.selectedOperationId = action.payload;
    },
    openPanel: (
      state,
      action: PayloadAction<{
        focusReturnElementId?: string;
        nodeId?: string;
        nodeIds?: string[];
        panelMode: PanelMode;
        referencePanelMode?: PanelMode;
      }>
    ) => {
      const { focusReturnElementId, nodeId, nodeIds, panelMode, referencePanelMode } = action.payload;
      const selectedNodes = nodeIds ? nodeIds : nodeId ? [nodeId] : [];

      clearPanel();

      state.currentPanelMode = panelMode;
      state.isCollapsed = false;
      state.focusReturnElementId = focusReturnElementId;
      state.previousPanelMode = referencePanelMode;
      state.connectionContent.selectedNodeIds = selectedNodes;
      state.discoveryContent.selectedNodeIds = selectedNodes;
      state.operationContent.selectedNodeId = selectedNodes[0];
    },
    setPinnedPanelActiveTab: (state, action: PayloadAction<string | undefined>) => {
      state.operationContent.pinnedNodeActiveTabId = action.payload;

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
    setSelectedPanelActiveTab: (state, action: PayloadAction<string | undefined>) => {
      state.errorContent.selectedTabId = action.payload;
      state.operationContent.selectedNodeActiveTabId = action.payload;

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
    setIsPanelLoading: (state, action: PayloadAction<boolean>) => {
      state.isLoading = action.payload;
    },
    setIsCreatingConnection: (state, action: PayloadAction<boolean>) => {
      state.connectionContent.isCreatingConnection = action.payload;
    },
    selectErrorsPanelTab: (state, action: PayloadAction<string>) => {
      state.errorContent.selectedTabId = action.payload;

      LoggerService().log({
        level: LogEntryLevel.Verbose,
        area,
        message: action.type,
        args: [action.payload],
      });
    },
  },
  extraReducers: (builder) => {
    builder.addCase(resetWorkflowState, () => initialState);

    // The below matchers are used to ensure that dispatches made to v1 panel are consumed by v2 panel as well.
    // Once v1 panel is deprecated & deleted, these can be safely removed.
    builder.addMatcher(isAnyOf(panelV1ExpandPanel), (state) => panelSlice.caseReducers.expandPanel(state));
    builder.addMatcher(isAnyOf(panelV1CollapsePanel), (state) => panelSlice.caseReducers.collapsePanel(state));
    builder.addMatcher(isAnyOf(panelV1ClearPanel), (state, action) =>
      panelSlice.caseReducers.clearPanel(state, { ...action, payload: { clearPinnedState: true } })
    );
    builder.addMatcher(isAnyOf(panelV1UpdatePanelLocation), (state, action) => panelSlice.caseReducers.updatePanelLocation(state, action));
    builder.addMatcher(isAnyOf(panelV1SetSelectedNodeId), (state, action) => panelSlice.caseReducers.setSelectedNodeId(state, action));
    builder.addMatcher(isAnyOf(panelV1SetSelectedNodeIds), (state, action) => panelSlice.caseReducers.setSelectedNodeIds(state, action));
    builder.addMatcher(isAnyOf(panelV1ChangePanelNode), (state, action) => panelSlice.caseReducers.changePanelNode(state, action));
    builder.addMatcher(isAnyOf(panelV1ExpandDiscoveryPanel), (state, action) =>
      panelSlice.caseReducers.expandDiscoveryPanel(state, action)
    );
    builder.addMatcher(isAnyOf(panelV1SelectOperationGroupId), (state, action) =>
      panelSlice.caseReducers.selectOperationGroupId(state, action)
    );
    builder.addMatcher(isAnyOf(panelV1SelectOperationId), (state, action) => panelSlice.caseReducers.selectOperationId(state, action));
    builder.addMatcher(isAnyOf(panelV1OpenPanel), (state, action) => panelSlice.caseReducers.openPanel(state, action));
    builder.addMatcher(isAnyOf(panelV1SelectPanelTab), (state, action) => panelSlice.caseReducers.setSelectedPanelActiveTab(state, action));
    builder.addMatcher(isAnyOf(panelV1SetIsPanelLoading), (state, action) => panelSlice.caseReducers.setIsPanelLoading(state, action));
    builder.addMatcher(isAnyOf(panelV1SetIsCreatingConnection), (state, action) =>
      panelSlice.caseReducers.setIsCreatingConnection(state, action)
    );
    builder.addMatcher(isAnyOf(panelV1SelectErrorsPanelTab), (state, action) =>
      panelSlice.caseReducers.selectErrorsPanelTab(state, action)
    );
    // End v1 panel matchers.
  },
});

export const {
  changePanelNode,
  clearPanel,
  collapsePanel,
  expandDiscoveryPanel,
  expandPanel,
  openPanel,
  selectErrorsPanelTab,
  selectOperationGroupId,
  selectOperationId,
  setPinnedPanelActiveTab,
  setSelectedPanelActiveTab,
  setIsCreatingConnection,
  setIsPanelLoading,
  setPinnedNode,
  setSelectedNodeId,
  setSelectedNodeIds,
  updatePanelLocation,
} = panelSlice.actions;

export default panelSlice.reducer;
