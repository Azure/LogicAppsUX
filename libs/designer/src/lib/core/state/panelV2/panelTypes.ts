import type { PanelLocation } from '@microsoft/designer-ui';

export interface PanelState {
  connectionContent: ConnectionPanelContentState;
  currentPanelMode: PanelMode;
  discoveryContent: DiscoveryPanelContentState;
  errorContent: ErrorPanelContentState;
  focusReturnElementId?: string;
  isCollapsed: boolean;
  isLoading: boolean;
  location: PanelLocation;
  nodeSearchContent: NodeSearchPanelContentState;
  operationContent: OperationPanelContentState;
  previousPanelMode: PanelMode | undefined;
  workflowParametersContent: WorkflowParametersPanelContentState;
}

export interface RelationshipIds {
  graphId: string;
  parentId?: string;
  childId?: string;
}

export interface ConnectionPanelContentState {
  isCreatingConnection: boolean;
  panelMode: 'Connection';
  selectedNodeIds: string[];
}

export interface DiscoveryPanelContentState {
  isAddingTrigger: boolean;
  isParallelBranch: boolean;
  panelMode: 'Discovery';
  relationshipIds: RelationshipIds;
  selectedNodeIds: string[];
  selectedOperationGroupId: string;
  selectedOperationId: string;
}

export interface ErrorPanelContentState {
  panelMode: 'Error';
  selectedTabId?: string;
}

export interface OperationPanelContentState {
  panelMode: 'Operation';
  pinnedNodeId?: string;
  pinnedNodeActiveTabId?: string;
  selectedNodeId?: string;
  selectedNodeActiveTabId?: string;
}

export interface NodeSearchPanelContentState {
  panelMode: 'NodeSearch';
}

export interface WorkflowParametersPanelContentState {
  panelMode: 'WorkflowParameters';
}

export type PanelContentState =
  | ConnectionPanelContentState
  | DiscoveryPanelContentState
  | ErrorPanelContentState
  | OperationPanelContentState
  | NodeSearchPanelContentState
  | WorkflowParametersPanelContentState;

export type PanelMode = PanelContentState['panelMode'];
