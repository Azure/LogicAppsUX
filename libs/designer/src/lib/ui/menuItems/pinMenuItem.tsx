import { useIsNodePinned } from '../../core/state/panelV2/panelSelectors';
import { MenuItem } from '@fluentui/react-components';
import { bundleIcon, Pin24Filled, Pin24Regular, PinOff24Filled, PinOff24Regular } from '@fluentui/react-icons';
import { LogEntryLevel, LoggerService } from '@microsoft/logic-apps-shared';
import { useCallback } from 'react';
import { useIntl } from 'react-intl';

const PinIcon = bundleIcon(Pin24Filled, Pin24Regular);
const UnpinIcon = bundleIcon(PinOff24Filled, PinOff24Regular);

export interface PinMenuItemProps {
  nodeId: string;
  onClick: (e: unknown) => void;
}

export const PinMenuItem: React.FC<PinMenuItemProps> = (props) => {
  const { nodeId, onClick } = props;

  const isNodePinned = useIsNodePinned(nodeId);

  const intl = useIntl();

  const pinAction = intl.formatMessage({
    defaultMessage: 'Pin Action',
    id: 'yy7jg4',
    description: 'Text indicating a menu button to pin an action to the side panel',
  });

  const unpinAction = intl.formatMessage({
    defaultMessage: 'Unpin Action',
    id: 'UFMpGk',
    description: 'Text indicating a menu button to unpin a pinned action from the side panel',
  });

  const handleClick = useCallback<PinMenuItemProps['onClick']>(
    (e) => {
      onClick(e);
      LoggerService().log({
        area: 'PinMenuItem:handleClick',
        args: [`isNodePinned:${isNodePinned}`],
        level: LogEntryLevel.Verbose,
        message: 'Action pin/unpin clicked.',
      });
    },
    [isNodePinned, onClick]
  );

  return (
    <MenuItem
      key={`PinMenuItem.${nodeId}`}
      icon={isNodePinned ? <UnpinIcon /> : <PinIcon />}
      onClick={handleClick}
      data-automation-id={'msla-pin-menu-option'}
    >
      {isNodePinned ? unpinAction : pinAction}
    </MenuItem>
  );
};
