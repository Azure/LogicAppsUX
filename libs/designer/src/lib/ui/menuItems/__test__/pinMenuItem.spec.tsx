import React from 'react';
import { PinMenuItem } from '../pinMenuItem';
import renderer from 'react-test-renderer';
import { describe, vi, beforeEach, afterEach, beforeAll, afterAll, it, test, expect } from 'vitest';

import * as PanelSelectors from '../../../core/state/panelV2/panelSelectors';

describe('lib/ui/menuItems/pinMenuItem', () => {
  it.each([true, false])('should render for actions if pinned=%s', (isPinned) => {
    vi.spyOn(PanelSelectors, 'useIsNodePinnedToOperationPanel').mockReturnValue(isPinned);
    const tree = renderer.create(<PinMenuItem nodeId="List_folder" onClick={vi.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
