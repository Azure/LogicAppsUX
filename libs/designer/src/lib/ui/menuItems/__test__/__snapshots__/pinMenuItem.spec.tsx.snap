// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lib/ui/menuItems/pinMenuItem > should render for actions if pinned=false 1`] = `
<div
  className="fui-MenuItem rszfvis"
  data-automation-id="msla-pin-menu-option"
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseEnter={[Function]}
  role="menuitem"
  tabIndex={0}
>
  <span
    className="fui-MenuItem__icon r9c34qo"
  >
    <svg
      aria-hidden={true}
      className="fui-Icon-filled ___1vjqft9_1uxjcqz fjseox fez10in fg4l7m0"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m21.07 7.76-4.83-4.83a2.75 2.75 0 0 0-4.4.72L9.4 8.52a.75.75 0 0 1-.42.37L4.8 10.33a1.25 1.25 0 0 0-.48 2.07l3.1 3.1L3 19.94V21h1.06l4.44-4.44 3.1 3.1c.66.66 1.77.4 2.07-.47l1.44-4.17c.06-.18.2-.33.37-.42l4.87-2.44a2.75 2.75 0 0 0 .72-4.4Z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden={true}
      className="fui-Icon-regular ___12fm75w_1vnroxd f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m16.24 2.93 4.83 4.83a2.75 2.75 0 0 1-.72 4.4l-4.87 2.44a.75.75 0 0 0-.37.42l-1.44 4.17a1.25 1.25 0 0 1-2.07.48l-3.1-3.1L4.06 21H3v-1.06l4.44-4.44-3.1-3.1c-.66-.66-.4-1.77.47-2.07l4.17-1.44c.18-.06.34-.2.42-.37l2.44-4.87a2.75 2.75 0 0 1 4.4-.72Zm3.77 5.89-4.83-4.83c-.6-.6-1.62-.44-2 .33l-2.44 4.87c-.26.52-.72.93-1.27 1.12l-3.8 1.3 6.71 6.71 1.31-3.79c.2-.55.6-1.01 1.12-1.27l4.87-2.44c.77-.38.93-1.4.33-2Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="fui-MenuItem__content r1ls86vo"
  >
    Pin Action
  </span>
</div>
`;

exports[`lib/ui/menuItems/pinMenuItem > should render for actions if pinned=true 1`] = `
<div
  className="fui-MenuItem rszfvis"
  data-automation-id="msla-pin-menu-option"
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseEnter={[Function]}
  role="menuitem"
  tabIndex={0}
>
  <span
    className="fui-MenuItem__icon r9c34qo"
  >
    <svg
      aria-hidden={true}
      className="fui-Icon-filled ___1vjqft9_1uxjcqz fjseox fez10in fg4l7m0"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.28 2.22a.75.75 0 0 0-1.06 1.06l5.9 5.9-3.3 1.15a1.25 1.25 0 0 0-.49 2.07l3.1 3.1L3 19.94V21h1.06l4.44-4.44 3.1 3.1c.66.66 1.77.4 2.07-.47l1.14-3.31 5.91 5.9a.75.75 0 0 0 1.06-1.06L3.28 2.22Zm17.07 9.94-3.34 1.67L10.17 7l1.67-3.34a2.75 2.75 0 0 1 4.4-.72l4.83 4.83a2.75 2.75 0 0 1-.72 4.4Z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden={true}
      className="fui-Icon-regular ___12fm75w_1vnroxd f1w7gpdv fez10in fg4l7m0"
      fill="currentColor"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.28 2.22a.75.75 0 0 0-1.06 1.06l5.9 5.9-3.3 1.15a1.25 1.25 0 0 0-.49 2.07l3.1 3.1L3 19.94V21h1.06l4.44-4.44 3.1 3.1c.66.66 1.77.4 2.07-.47l1.14-3.31 5.91 5.9a.75.75 0 0 0 1.06-1.06L3.28 2.22ZM13.64 14.7l-1.26 3.62-6.7-6.7 3.62-1.26 4.34 4.34Zm6.04-3.88-3.78 1.9L17 13.82l3.34-1.67a2.75 2.75 0 0 0 .72-4.4l-4.83-4.83a2.75 2.75 0 0 0-4.4.72l-1.67 3.34 1.12 1.11 1.89-3.78c.38-.77 1.4-.93 2-.33l4.83 4.83c.6.6.44 1.62-.33 2Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="fui-MenuItem__content r1ls86vo"
  >
    Unpin Action
  </span>
</div>
`;
